Triggering Endings for Stats Hitting Zero

motzca opened the pull request.

motzca committed:
Adding EndingManager + Starting to incorporate it into the GameManager

motzca committed:
Changing to cardId instead of cardName

motzca committed:
Temp push

motzca committed:
temp push

motzca committed:
Pushing for new zero status endings

motzca committed:
Merge branch 'master' into ZeroStatEndings

motzca committed:
Merge branch 'master' into ZeroStatEndings

motzca committed:
Adding StatusChecker - still needs to be tested with the specialty cards

LostInTyping commented on Assets/Scripts/Game/EndingManager.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        if (GameManager.MoneyStatus <= 0 && OnMoneyZero != null)
+        {
+            OnMoneyZero(endingMoneyCardId);
```
In the current implementation, you manually check for null before invoking events, which, while functional, adds unnecessary lines to our codebase. I recommend adopting the null conditional operator (?.) for event invocation, like this:
```
if (GameManager.MoneyStatus <= 0)
{
    OnMoneyZero?.Invoke(endingMoneyCardId);
}
```
I think it enhances the readability to the code, improves the safety because using ?, we automatically bypass the event invocation if it's null, thus eliminating the risk of encountering a NullReferenceException, and lastly using the conditional operator aligns our code with current best practices.

Do this for all 3

LostInTyping commented on Assets/Scripts/Game/EndingManager.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            OnMoneyZero(endingMoneyCardId);
+        }
+        else if (GameManager.EnergyStatus <= 0 && OnEnergyZero != null)
```
To better manage game stat changes, we should utilize individual `if` statements for each condition. This approach allows for simultaneous triggering of multiple events, which I think will enhances the clarity of the code, and guarantees a more precise and responsive game state management. In the current implementation I believe this is fine but we can have more varied endings easier if we watch all stats in the future but not a big deal either way currently.

LostInTyping commented on Assets/Scripts/Game/StatusChecker.cs:
```diff
@@ -0,0 +1,70 @@
+        resourceManager = FindObjectOfType<ResourceManager>();
+    }
+
+    void Update()
```
Instead oh having everything in update maybe we should introduced a new method and call it CheckForStatusZero to encapsulate status checks, keeping Update clean and focused on just calling the method.

so do somthing like this? 

```  
void Update()
    {
        CheckForStatusZero();
    }

    private void CheckForStatusZero()
    {
        if (GameManager.MoneyStatus <= 0)
        {
            LoadSpecialEndingCard(moneyZeroCard);
            return;
        }

        if (GameManager.EnergyStatus <= 0)
        {
            LoadSpecialEndingCard(energyZeroCard);
            return;
        }

        if (GameManager.ReputationStatus <= 0)
        {
            LoadSpecialEndingCard(reputationZeroCard);
            return;
        }
    }
```

LostInTyping commented on Assets/Scripts/Game/GameManager.cs:
```diff
@@ -67,6 +74,19 @@ void Update()
         HandleCardInput();
         UpdateDialogue();
+
+        if (MoneyStatus <= 0 && OnMoneyZero != null)
```
same as in EndingManager.cs what if we do: 

```
if (GameManager.MoneyStatus <= 0)
{
    OnMoneyZero?.Invoke(endingMoneyCardId);
}

if (GameManager.EnergyStatus <= 0)
{
    OnEnergyZero?.Invoke(endingEnergyCardId);
}

if (GameManager.ReputationStatus <= 0)
{
    OnReputationZero?.Invoke(endingReputationCardId);
}
```

LostInTyping reviewed (changes requested):
In addition on line 132 of gameManager.cs we could do the following:

```
public void UpdateMoney(int change)
{
    int newMoneyStatus = Mathf.Clamp(MoneyStatus + change, MinValue, MaxValue);
    if (MoneyStatus != newMoneyStatus)
    {
        MoneyStatus = newMoneyStatus;
        OnMoneyZero?.Invoke(endingMoneyCardId); 
    }
}
```

Updating `UpdateMoney`, `UpdateEnergy`, and `UpdateReputation` methods using the above as an example provides several enhancements that collectively improve the game's logic and code quality. This ensures that updates and event triggers only occur following actual changes, eliminating unnecessary operations that could worsen the game's performance. I also think the use of the null-conditional operator for event invocation simplifies the code by removing the need for explicit null checks, streamlining event handling.

motzca committed:
Merge branch 'master' into ZeroStatEndings

motzca committed:
Added ending blocks and updated ending manager - still working on triggering the game to end and return back to the start emn

motzca committed:
Reverting initial changes - need to rework for using Fungus

motzca committed:
Endings now triggering - still need to work on getting the game to reset

motzca committed:
Temp push

motzca committed:
Merge branch 'master' into ZeroStatEndings

motzca committed:
Can restart the game and go back to start menu

motzca commented:
@L0ST201 Just to let you know, when resetting back to the beginning, it takes a few seconds to restart the full game. If needed, I can try to modify that code so that it runs smoother.

motzca committed:
Update Day1.unity

LostInTyping commented:
There are a few problems. 

- [x] 1. Currently the game does not Restart it sends errors this looks like you have it targeting scene `MainMenu` When this doesn't exist and is just `Start`.
- [x] 2. The game on Restart doesn't Currently Reset stats and keeps them at there current state when they should reset to default.
![image](https://github.com/motzca/SeniorDesignProject_AnimalVilla/assets/82691998/6a9a9a65-f7ec-4477-8269-98e5db974ed2)
![image](https://github.com/motzca/SeniorDesignProject_AnimalVilla/assets/82691998/1254481e-ee43-4518-b7a2-5eee6449cc99)

- [x] Text falls outside of text box.

LostInTyping committed:
Card Animation Changes

1. Increased the distance to show the continue text.
2. Increased the distance for card swiping.
3. The card slides back smoothly on a failed swipe and is improved from its original behavior.
4. Added a new card animation for successful swipes, and the cards reappear in its original position.

LostInTyping committed:
Update CardController.cs

- Removed HandleMouseInputs as this is no longer needed.
- Added effect to return card to fade back to full opacity on failed swiped.

LostInTyping committed:
card animation

Card animations + Static card "to pull card from"

motzca committed:
Update based on code review feedback

motzca committed:
Variables reset to 50  when resetting game

motzca committed:
Making sure that all variables don't increase/decrease when reaching an ending

motzca committed:
Merge branch 'master' into ZeroStatEndings

motzca committed:
Status variables reset whenever we start a new game
